src/wallet/wallet.cpp

```diff
@@ -4002,6 +4002,23 @@ bool CWallet::ApplyMigrationData(MigrationData& data, bilingual_str& error)
             }
         }
     }
+
+    // Persist added address book entries (labels, purpose) for watchonly and solvable wallets
+    auto persist_address_book = [](const CWallet& wallet) {
+        LOCK(wallet.cs_wallet);
+        WalletBatch batch{wallet.GetDatabase()};
+        for (const auto& [destination, addr_book_data] : wallet.m_address_book) {
+            auto address{EncodeDestination(destination)};
+            auto purpose{addr_book_data.purpose};
+            auto label{addr_book_data.GetLabel()};
+            // don't bother writing default values (unknown purpose, empty label)
+            if (purpose != "unknown") batch.WritePurpose(address, purpose);
+            if (!label.empty()) batch.WriteName(address, label);
```

Comment:
In commit "wallet: fully migrate address book entries for watchonly/solvable wallets" (d5f4ae7fac0bceb0c9ad939b9a4fbdb85da0bf95)

It seems like a bug to not write the label when then label is empty, because will make receiving address with empty labels labels appear to be change addresses. If the label isn't written `CAddressBookData::SetLabel` will not be called:

https://github.com/bitcoin/bitcoin/blob/f50fb178c30ea4bec0b369af3d15cab08d33396f/src/wallet/wallet.h#L219-L222

so `CAddressBookData::IsChange` will return true

https://github.com/bitcoin/bitcoin/blob/f50fb178c30ea4bec0b369af3d15cab08d33396f/src/wallet/wallet.h#L217

Probably this line should say:

```c++
if (!addr_book_data.IsChange()) batch.WriteName(address, addr_book_data.GetLabel());
```

Reply:
Check #26836, bug should be fixed there. Have re-written the process there without the `persist_address_book` calls.

Reply:
Thanks will check it out. Also would be good to have a test for the bug.

Reply:
yeah 👌🏼.